pyfr/backends/base/types.py

```diff
@@ -16,34 +16,46 @@ def __init__(self, backend, dtype, ioshape, initval, extent, aliases,
         self.dtype = dtype
         self.itemsize = np.dtype(dtype).itemsize
 
-        # Alignment requirement for the leading dimension
-        ldmod = backend.alignb // self.itemsize if 'align' in tags else 1
-
         # Our shape and dimensionality
         shape, ndim = list(ioshape), len(ioshape)
 
+        # SoA and block column size
+        soasz, csubsz = backend.soasz, backend.csubsz
+
+        self.blocked = backend.blocks and 'opmat' not in self.tags
+        self.aligned = 'align' in self.tags
+
         if ndim == 2:
             nrow, ncol = shape
             aosoashape = shape
+
+            # Alignment requirement for the leading dimension
+            ldmod = csubsz if self.aligned else 1
+            nbcol = csubsz if self.blocked else ncol - (ncol % -ldmod)
         else:
-            nvar, narr, k = shape[-2], shape[-1], backend.soasz
-            nparr = narr - narr % -k
+            nvar, narr, k = shape[-2], shape[-1], soasz
+            nparr = narr - narr % -csubsz
 
             nrow = shape[0] if ndim == 3 else shape[0]*shape[1]
             ncol = nvar*nparr
             aosoashape = shape[:-2] + [nparr // k, nvar, k]
+            nbcol = nvar*csubsz if self.blocked else ncol
 
         # Assign
-        self.nrow, self.ncol = int(nrow), int(ncol)
+        self.nrow, self.ncol, self.nbcol = int(nrow), int(ncol), int(nbcol)
 
         self.datashape = aosoashape
         self.ioshape = ioshape
 
-        self.leaddim = self.ncol - (self.ncol % -ldmod)
+        self.splitsz = self.nbcol if backend.blocks else soasz
+        self.blocksz = self.nrow*self.nbcol
+        self.nblocks = (self.ncol - self.ncol % -self.nbcol) // self.nbcol
+        self.nbytes = self.nblocks*self.blocksz*self.itemsize
 
-        self.pitch = self.leaddim*self.itemsize
-        self.nbytes = self.nrow*self.pitch
-        self.traits = (self.nrow, self.ncol, self.leaddim, self.dtype)
+        self.leaddim = self.nbcol
```

Comment:
What is the purpose of leaddim (or nbcol), seems redundant.

Reply:
I thought using leaddim argument as number of columns in a block (nbcol) would be confusing. Because leaddim somehow indicates a big number related to the number of elements and variables, but with cache blocking it is the number of columns in a block. If you think this is bad practice or confusing I can use leaddim only whenever I need.